### PR TITLE
zpool import -d to specify device path

### PR DESCRIPTION
**Pull request by tuxoko**
When we know which devices have the pool we are looking for, sometime
it's better if we can directly pass those device paths to zpool import
instead of letting it to search through all unrelated stuff, which might
take a lot of time if you have hundreds of disks.

This patch allows option -d <dev_path> to zpool import. You can have
multiple pairs of -d <dev_path>, and zpool import will only search
through those devices. For example:

    zpool import -d /dev/sda -d /dev/sdb

Signed-off-by: Chunwei Chen <david.chen@osnexus.com>

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux code style requirements.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] All commit messages are properly formatted and contain `Signed-off-by`.
- [ ] Change has been approved by a ZFS on Linux member.

### CODE REVIEW

**Comment by tuxoko** [comment]
Followup from #6423

**Comment by tonyhutter** [comment]
> You can have multiple pairs of -d <dev_path>, and zpool import will only search through those devices.

I'd recommend a single `-d` followed by all the devices/paths. That way you can just paste in the list of paths.   See my earlier comment (https://github.com/zfsonlinux/zfs/pull/6423#issuecomment-347035949)

**Comment by tuxoko** [comment]
@tonyhutter 
Currently, -d only allows 1 argument. Are you saying we change it all together?

**Comment by tuxoko** [comment]
Wouldn't it cause ambiguity on whether the last argument is a -d argument or pool name.

**Comment by tonyhutter** [comment]
> Wouldn't it cause ambiguity on whether the last argument is a -d argument or pool name.
> Currently, -d only allows 1 argument. Are you saying we change it all together?

Yep, that's what I'm saying.  I see your problem with it though:

```
$ mkdir tank
$ truncate -s 100M tank/file
$ sudo zpool create tank `pwd`/tank/file
$ sudo zpool export tank 
$ sudo zpool import tank
$ sudo zpool import -d tank
   pool: tank
     id: 17116745296311198214
  state: ONLINE
 action: The pool can be imported using its name or numeric identifier.
 config:

	tank                      ONLINE
	  /home/hutter/tank/file  ONLINE
$ sudo zpool import -d tank tank                    <<<< this stops working
```
You could get around this by requiring the user to put the path list in quotes if there's more than one path:
`zpool import -d '/dev/sda /dev/sdb /dev/disk/by-vdev' poolname`

That said, I was surprised to see that the current `zpool import` man page in master already says:

`           -d dir          Searches for devices or files in dir. The -d option can be specified multiple times.`

So your patch would be fixing `zpool import` to actually do what the man page says it should doing already (which is a good thing).  

@behlendorf any thoughts?

File: lib/libzfs/libzfs_import.c [hunk 1]
```diff
@@ -1676,6 +1677,120 @@ zpool_clear_label(int fd)
...
+		}
+
+		zfs_error_aux(hdl, strerror(error));
+		(void) zfs_error_fmt(hdl, EZFS_BADPATH, dgettext(
+		    TEXT_DOMAIN, "cannot resolve path '%s'"), dir);
+		goto out;
+	}
+
+	zpool_find_import_scan_add_slice(hdl, lock, cache, path, name,
+	    order);
```
Comment (behlendorf) [inline lib/libzfs/libzfs_import.c]: [nit] This does fit on one line and there's no need to wrap it.  You can also drop the extra line of whitespace after this one.

File: lib/libzfs/libzfs_import.c [hunk 2]
```diff
@@ -1687,18 +1802,16 @@ zpool_find_import_scan(libzfs_handle_t *hdl, pthread_mutex_t *lock,
 	rdsk_node_t *slice;
 	void *cookie;
 	int i, error;
+	struct stat sbuf;
```
Comment (behlendorf) [inline lib/libzfs/libzfs_import.c]: You might want to move this declaration in to the `for ()` block to make the scope clearer.

File: lib/libzfs/libzfs_import.c [hunk 3]
```diff
@@ -1676,6 +1677,120 @@ zpool_clear_label(int fd)
...
+	if (asprintf(&slice->rn_name, "%s/%s", path, name) == -1) {
+		free(slice);
+		return;
+	}
+	slice->rn_vdev_guid = 0;
+	slice->rn_lock = lock;
+	slice->rn_avl = cache;
+	slice->rn_hdl = hdl;
+	slice->rn_order = order + IMPORT_ORDER_SCAN_OFFSET;
+	slice->rn_labelpaths = B_FALSE;
```
Comment (behlendorf) [inline lib/libzfs/libzfs_import.c]: [nit] I thinking adding an empty line before the `pthread_mutex_lock` call would improve readability.

**Comment by tuxoko** [comment]
Updated according to review.

File: tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+# CDDL HEADER END
+#
+
+#
+# Copyright 2007 Sun Microsystems, Inc.  All rights reserved.
+# Use is subject to license terms.
+#
+
+#
+# Copyright (c) 2012, 2016 by Delphix. All rights reserved.
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh]: You'll want to use the short CDDL header for new tests and update the copyright assignment accordingly.  For example:

```
#
# This file and its contents are supplied under the terms of the
# Common Development and Distribution License ("CDDL"), version 1.0.
# You may only use this file in accordance with the terms of version
# 1.0 of the CDDL.
#
# A full copy of the text of the CDDL should have accompanied this
# source.  A copy of the CDDL is also available via the Internet at
# http://www.illumos.org/license/CDDL.
#

#
# Copyright (c) 2018 by <your name/company here>
#
```

File: tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#
+
+. $STF_SUITE/include/libtest.shlib
+. $STF_SUITE/tests/functional/cli_root/zpool_import/zpool_import.cfg
+
+#
+# DESCRIPTION:
+#	Destroyed pools are not listed unless with -D option is specified.
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh]: Description needs to be updated.

File: tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+#
+
+verify_runnable "global"
+
+function cleanup
+{
+	destroy_pool $TESTPOOL1
+
+	log_must rm $VDEV0 $VDEV1
+	log_must mkfile $FILE_SIZE $VDEV0 $VDEV1
```
Comment (behlendorf) [inline tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh]: Might as well use `truncate -s` here to speed things up.

File: tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+log_assert "Pool can be imported with '-d <device>'"
+log_onexit cleanup
+
+log_must zpool create $TESTPOOL1 $VDEV0 $VDEV1
+log_must zpool export $TESTPOOL1
+log_must zpool import -d $VDEV0 -d $VDEV1 $TESTPOOL1
```
Comment (tonyhutter) [inline tests/zfs-tests/tests/functional/cli_root/zpool_import/zpool_import_015_pos.ksh]: Can you put `$VDEV1` in a separate directory, just so we can test both `-d file` and `-d dir`.

**Comment by tuxoko** [comment]
Updated according to review.

**Pull request merged**